File: Python/700 Search in a Binary Search Tree.py
# Search in a Binary Search Tree
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def searchBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        if not root:
            return []
        queue = []
        queue.append(root)
        while queue:
            curr = queue.pop(0)
            if curr.val == val:
                return curr
            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)
        return []

File: C++/700 Search in a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        queue<TreeNode*> q;
        if (root != nullptr) q.push(root);
        else return root;
        while (!q.empty()) {
            TreeNode* tmp = q.front();
            q.pop();
            if (tmp->val == val) return tmp;
            else if (tmp->val > val && tmp->left != nullptr) q.push(tmp->left);
            else if (tmp->right!=nullptr) q.push(tmp->right);
        }
        return nullptr;
        // while (root!=nullptr) {
        //     if (root->val == val) return root;
        //     else if (val < root->val) root = root->left;
        //     else root = root->right;
        // }
        // return root;
    }
};